drivers/gpio/gpio_pca_series.c

```diff
@@ -0,0 +1,2169 @@
+		}
+	}
+
+	if (flags & GPIO_INPUT) {
+		/* configure PCA_REG_TYPE_1B_CONFIGURATION */
+		ret = gpio_pca_series_reg_read_cache(dev,
+			PCA_REG_TYPE_1B_CONFIGURATION, (uint8_t *)&reg_value);
+		reg_value |= (BIT(pin)); /* set bit to set input */
+		ret = gpio_pca_series_reg_write(dev,
+			PCA_REG_TYPE_1B_CONFIGURATION, (uint8_t *)&reg_value);
+	} else {
+		uint32_t out_old, out;
+#ifdef CONFIG_GPIO_PCA_SERIES_CACHE_ALL
+		/* get output register old value from reg cache */
+		ret = gpio_pca_series_reg_read_cache(dev, PCA_REG_TYPE_1B_OUTPUT_PORT,
+							 (uint8_t *)&out_old);
+		if (ret) {
+			ret = -EINVAL; /* should never fail */
+			goto out;
+		}
+#else  /* CONFIG_GPIO_PCA_SERIES_CACHE_ALL */
+		out_old = gpio_pca_series_reg_cache_mini_get(dev)->output;
+#endif /* CONFIG_GPIO_PCA_SERIES_CACHE_ALL */
+
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			out = out_old | BIT(pin);
+			ret = gpio_pca_series_reg_write(dev, PCA_REG_TYPE_1B_OUTPUT_PORT,
+							(uint8_t *)&out);
+		}
+		if (flags & GPIO_OUTPUT_INIT_LOW) {
+			out = out_old & (~BIT(pin));
+			ret = gpio_pca_series_reg_write(dev, PCA_REG_TYPE_1B_OUTPUT_PORT,
+							(uint8_t *)&out);
+		}
+		if (ret == 0) {
+#ifndef CONFIG_GPIO_PCA_SERIES_CACHE_ALL
+			/** update output register old value to void* cache raw value */
+			gpio_pca_series_reg_cache_mini_get(dev)->output = out;
+#endif /* CONFIG_GPIO_PCA_SERIES_CACHE_ALL */
+		}
+		/* configure PCA_REG_TYPE_1B_CONFIGURATION */
+		ret = gpio_pca_series_reg_read_cache(dev, PCA_REG_TYPE_1B_CONFIGURATION,
+							 (uint8_t *)&reg_value);
+		reg_value &= (~BIT(pin)); /* clear bit to set output */
+		ret = gpio_pca_series_reg_write(dev, PCA_REG_TYPE_1B_CONFIGURATION,
+						(uint8_t *)&reg_value);
+	}
+
+out:
+	k_sem_give(&data->lock);
+	LOG_DBG("dev %s configure return %d", dev->name, ret);
+	return ret;
+}
+
+/**
+ * @brief read gpio port
+ *
+ * @note read input_port register will clear interrupt masks
+ *       on supported devices. This API is used for part no
+ *       without PCA_HAS_INT_EXTEND capability.
+ *
+ * @param dev
+ * @param value
+ * @return int 0            if success
+ *             -EWOULDBLOCK if called from ISR context
+ */
+static int gpio_pca_series_port_read_standard(
+		const struct device *dev, gpio_port_value_t *value)
+{
+	struct gpio_pca_series_data *data = dev->data;
+	uint32_t input_data;
+	int ret = 0;
+
+	/* Can't do I2C bus operations from an ISR */
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	LOG_DBG("dev %s standard_read", dev->name);
+	k_sem_take(&data->lock, K_FOREVER);
+
+	/* Read Input Register */
+	ret = gpio_pca_series_reg_read(dev,
+		PCA_REG_TYPE_1B_INPUT_PORT, (uint8_t *)&input_data);
+	if (ret) {
+		goto out;
+	}
+
+	*value = input_data;
+#ifdef CONFIG_GPIO_PCA_SERIES_INTERRUPT
+#ifndef CONFIG_GPIO_PCA_SERIES_CACHE_ALL
+	/** cache input for interrupt handling */
+	gpio_pca_series_reg_cache_mini_get(dev)->input = input_data;
+	/* FIXME: handle interrupt here, fire callbacks as needed */
```

Comment:
Need to address this.
For device w/o `PCA_HAS_INT_EXTEND` feature, an gpio read also clears interrupt flag. So after reading, driver needs to check if interrupt is triggered and fire callbacks.

Reply:
Based on this comment I assume you still  have some debug work planned before this PR is ready to merge. Could you do one of the following, so I know when this PR is ready for a final review?
- mark this PR as draft, and move it out of draft status when it is ready
- ping me explicitly when the PR is tested and ready to merge from your point of view

Reply:
sure. should be ready next week.

Reply:
be aware that Zephyr's next release has an RC1 feature freeze at the end of next week. You should strive to have this ready before this if you want this part of the next release

Reply:
Uhh...Just a update here...

I got some other work to do and just came back. I'm still working on this pr. I tested PCAL6524 on i.MX93 EVK, it's working fine w/o cache enabled. input, output, interrupt tested with `sample/basic/button` app.

Cache API, a little bit complicated...

Since PCAL6524 has all the registers, I'll be able to simulate other devices. So still need more testing.